libutils/stack.c

```diff
@@ -0,0 +1,267 @@
+            GetErrorStrFromCode(ret));
+        pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_NORMAL);
+    }
+
+    stack->lock = malloc(sizeof(pthread_mutex_t));
+    ret = pthread_mutex_init(stack->lock, &attr);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to initialize mutex (pthread_mutex_init: %s)",
+            GetErrorStrFromCode(ret));
+        free(stack);
+        return NULL;
+    }
+
+    pthread_mutexattr_destroy(&attr);
+
+    stack->capacity = initial_capacity;
+    stack->size = STACK_EMPTY;
+    stack->data = xcalloc(initial_capacity, sizeof(void *));
+    stack->ItemDestroy = ItemDestroy;
+
+    return stack;
+}
+
+void StackDestroy(Stack *stack)
+{
+    if (stack != NULL)
+    {
+        ThreadLock(stack->lock);
+        DestroyRange(stack, STACK_EMPTY, stack->size);
+        ThreadUnlock(stack->lock);
+
+        StackSoftDestroy(stack);
+    }
+}
+
+void StackSoftDestroy(Stack *stack)
+{
+    if (stack != NULL)
+    {
+        if (stack->lock != NULL)
+        {
+            pthread_mutex_destroy(stack->lock);
+            free(stack->lock);
+        }
+
+        free(stack->data);
+        free(stack);
+    }
+}
+
+void *StackPop(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+
+    size_t size = stack->size;
+    void *item = NULL;
+
+    if (size > STACK_EMPTY)
+    {
+        size--;
+        item = stack->data[size];
+
+        stack->data[size] = NULL;
+        stack->size = size;
+    }
+
+    ThreadUnlock(stack->lock);
+
+    return item;
+}
+
+void StackPush(Stack *stack, void *item)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+
+    ExpandIfNecessary(stack);
+    stack->data[stack->size++] = item;
+
+    ThreadUnlock(stack->lock);
+}
+
+size_t StackPushReportCount(Stack *stack, void *item)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+
+    ExpandIfNecessary(stack);
+    stack->data[stack->size++] = item;
+    size_t size = stack->size;
+
+    ThreadUnlock(stack->lock);
+
+    return size;
+}
+
+size_t StackCount(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+    size_t count = stack->size;
+    ThreadUnlock(stack->lock);
+
+    return count;
+}
+
+size_t StackCapacity(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+    size_t capacity = stack->capacity;
+    ThreadUnlock(stack->lock);
+
+    return capacity;
+}
+
+bool StackIsEmpty(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+    bool const empty = (stack->size == STACK_EMPTY);
+    ThreadUnlock(stack->lock);
+
+    return empty;
+}
+
+Stack *StackCopy(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+
+    Stack *new_stack = StackNew(stack->capacity, stack->ItemDestroy);
+    for (size_t i = 0; i < stack->size; i++)
+    {
+        new_stack->data[i] = stack->data[i];
+    }
```

Comment:
So `new_stack->data = xmemdup(stack->data, sizeof(void *) * stack->size);`?
I have no way of knowing the size of the data elements themselves without supplying a "size" function.

Reply:
Yes, and of course not using `StackNew()`. I think this should still be a shallow copy, but the implementation should be more optimal than what a code without the knowledge of the internals could do (i.e. `xmemdup()`).

Reply:
Since you need a new mutex lock anyway, wouldn't it make more sense to just create a new object? I feel like it would lead to less duplicated code as well if you do that.

Reply:
It's not such a big amount of code duplicated. And it may provide some nice performance benefits.